### DISCUSSION TITLE
Limit max threads available to OpenVDB / TBB

### DISCUSSION THREAD

**Discussion by kdmarrett** [AcademySoftwareFoundation/openvdb#1615] [Q&A] [answered]
For a C++ library that uses OpenVDB as a dependency what is your recommended way of limiting the total # of threads that OpenVDB / TBB can spawn. I need a runtime user defined way to limit the max number of threads.

I'm not asking how to set the max # of threads for a given system .. that is already TBB's default.
I'd rather have a global solution (set at the begining of execution and forget) rather than having to wrap each openvdb call in a task arena. Note that tbb::task_scheduler_init is deprecated in recent TBB releases

Whats the recommended strategy?

I appreciate the help!

**Comment by Idclip** [comment, marked as answer]
You can use `tbb/global_control` to achieve this. Note though that this will affect all tbb execution in your application that links against the specified version from which the `::global_control` symbol is called. Here's an example that we use in the `vdb_ax` binary:

https://github.com/AcademySoftwareFoundation/openvdb/blob/e7ec2a716eac2bd03fc491d2d9639c6d5afd61cf/openvdb_cmd/vdb_ax/main.cc#L782-L786

If you're using other libraries which also use tbb but _only_ want to control tbb utilisation in VDB, unfortunately the only** way is to wrap each call in an arena, as you've mentioned.

** you could also use multiple namespaced versions of tbb but this is probably a bit much
